internal/resolver/dns/dns_resolver_test.go

```diff
@@ -16,1510 +16,865 @@
-		}
-		if !reflect.DeepEqual(v.want, state.Addresses) {
-			t.Errorf("Resolved addresses of target: %q = %+v, want %+v", v.target, state.Addresses, v.want)
-		}
-		r.ResolveNow(resolver.ResolveNowOptions{})
-		for i := 0; i < 50; i++ {
-			state, cnt = cc.getState()
-			if cnt > 1 {
-				t.Fatalf("Unexpected second call by resolver to UpdateState.  state: %v", state)
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			overrideResolutionRate(t, 0)
+			overrideTimeAfterFunc(t, 2*defaultTestTimeout)
+			r, stateCh, _ := buildResolverWithTestClientConn(t, test.target)
+
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+			defer cancel()
+			verifyUpdateFromResolver(ctx, t, stateCh, test.wantAddr, nil, "")
+
+			// Attempt to re-resolve should not result in a state update.
+			r.ResolveNow(resolver.ResolveNowOptions{})
+			sCtx, sCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
+			defer sCancel()
+			select {
+			case <-sCtx.Done():
+			case s := <-stateCh:
+				t.Fatalf("Unexpected state update from the resolver: %+v", s)
 			}
-			time.Sleep(time.Millisecond)
-		}
-		r.Close()
+		})
 	}
 }
 
-func TestResolveFunc(t *testing.T) {
-	defer leakcheck.Check(t)
-	defer func(nt func(d time.Duration) *time.Timer) {
-		newTimer = nt
-	}(newTimer)
-	newTimer = func(d time.Duration) *time.Timer {
-		// Will never fire on its own, will protect from triggering exponential backoff.
-		return time.NewTimer(time.Hour)
-	}
+// Tests the DNS resolver builder with different target names.
+func (s) TestResolverBuild(t *testing.T) {
 	tests := []struct {
-		addr string
-		want error
+		name    string
+		target  string
+		wantErr string
 	}{
 		// TODO(yuxuanli): More false cases?
-		{"www.google.com", nil},
-		{"foo.bar:12345", nil},
-		{"127.0.0.1", nil},
-		{"::", nil},
-		{"127.0.0.1:12345", nil},
-		{"[::1]:80", nil},
-		{"[2001:db8:a0b:12f0::1]:21", nil},
-		{":80", nil},
-		{"127.0.0...1:12345", nil},
-		{"[fe80::1%25lo0]:80", nil},
-		{"golang.org:http", nil},
-		{"[2001:db8::1]:http", nil},
-		{"[2001:db8::1]:", errEndsWithColon},
-		{":", errEndsWithColon},
-		{"", errMissingAddr},
-		{"[2001:db8:a0b:12f0::1", fmt.Errorf("invalid target address [2001:db8:a0b:12f0::1, error info: address [2001:db8:a0b:12f0::1:443: missing ']' in address")},
+		{
+			name:   "valid url",
+			target: "www.google.com",
+		},
+		{
+			name:   "host port",
+			target: "foo.bar:12345",
+		},
+		{
+			name:   "ipv4 address with default port",
+			target: "127.0.0.1",
+		},
+		{
+			name:   "ipv6 address without brackets and default port",
+			target: "::",
+		},
+		{
+			name:   "ipv4 address with non-default port",
+			target: "127.0.0.1:12345",
+		},
+		{
+			name:   "localhost ipv6 with brackets",
+			target: "[::1]:80",
+		},
+		{
+			name:   "ipv6 address with brackets",
+			target: "[2001:db8:a0b:12f0::1]:21",
+		},
+		{
+			name:   "empty host with port",
+			target: ":80",
+		},
+		{
+			name:   "",
```

Comment:
Add a name?

Reply:
I honestly don't know what this case is testing. Do you?

Reply:
Done.